### PR TITLE
customize for announcement

### PR DESCRIPTION
**Pull request by m-sugawara**
## Overview
 - add new cell for showing announcement.
 - remove some codes for adjusting layouts

## Screenshot
Simulator Screen Shot - iPhone X - 2019-08-29 at 14 59 29

@quipper/aya-native-ios-devs 
Please review if you have time. 🙏  I will merge this a few hours later.

### CODE REVIEW

File: Example/Sources/View Controllers/AdvancedExampleViewController.swift [hunk 1]
```diff
@@ -318,14 +318,15 @@ extension AdvancedExampleViewController: MessagesDisplayDelegate {
...
-        }
+
+        return .none
+//        return .custom { view in
+//            let radius: CGFloat = 16
+//            let path = UIBezierPath(roundedRect: view.bounds, byRoundingCorners: corners, cornerRadii: CGSize(width: radius, height: radius))
+//            let mask = CAShapeLayer()
+//            mask.path = path.cgPath
+//            view.layer.mask = mask
+//        }
```
Comment (manicmaniac) [inline Example/Sources/View Controllers/AdvancedExampleViewController.swift]: Could you please remove those comments?

File: Example/Sources/View Controllers/AnnouncementExampleViewController.swift [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+
+class AnnouncementExampleViewController: MessagesViewController {
+    override var preferredStatusBarStyle: UIStatusBarStyle {
+        return .lightContent
+    }
+
+    var conversationMessages: [ConversationMessage] = []
+
+    private var _dateFormatter: DateFormatter?
+    var dateFormatter: DateFormatter {
```
Comment (manicmaniac) [inline Example/Sources/View Controllers/AnnouncementExampleViewController.swift]: Why not `lazy var`?

File: Example/Sources/View Controllers/AnnouncementExampleViewController.swift [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+        }
+    }
+
+}
+
+extension AnnouncementExampleViewController: MessageCellDelegate {
+    func didTapMessage(in cell: MessageCollectionViewCell) {
+        guard let indexPath = messagesCollectionView.indexPath(for: cell),
+            let message = messagesCollectionView.messagesDataSource?.messageForItem(at: indexPath, in: messagesCollectionView) else { return }
+        print("didTapMessage \(message.messageId)")
```
Comment (manicmaniac) [inline Example/Sources/View Controllers/AnnouncementExampleViewController.swift]: Still need debug prints?

File: Example/Sources/View Controllers/AnnouncementExampleViewController.swift [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+    }
+
+    func didSelectURL(_ url: URL) {
+        print("url selected")
```
Comment (manicmaniac) [inline Example/Sources/View Controllers/AnnouncementExampleViewController.swift]: :eyes:

File: Example/Sources/View Controllers/AnnouncementExampleViewController.swift [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+        ]
+    }
+}
+
+struct ConversationMessage {
+    var sender: User
+    var text: String?
+    var image: UIImage?
+    var createdTs: TimeInterval
+    var isAnnouncement: Bool = false
```
Comment (manicmaniac) [inline Example/Sources/View Controllers/AnnouncementExampleViewController.swift]: Couldn't you declare those members as `let`?

File: Example/Sources/View Controllers/AnnouncementExampleViewController.swift [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+        self.createdTs = createdTs
+        self.isAnnouncement = isAnnouncement
+    }
+}
+
+struct MediaItemImpl: MediaItem {
+    var url: URL?
+    var image: UIImage?
+    var placeholderImage: UIImage
+    var size: CGSize
```
Comment (manicmaniac) [inline Example/Sources/View Controllers/AnnouncementExampleViewController.swift]: :eyes: `let`

File: Example/Sources/View Controllers/AnnouncementExampleViewController.swift [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+    static func other() -> User {
+        return User(senderId: "other", displayName: "other user")
+    }
+}
+
+struct MessageDisplaySource: MessageType {
+    var sender: SenderType
+    var messageId: String
+    var sentDate: Date
+    var kind: MessageKind
```
Comment (manicmaniac) [inline Example/Sources/View Controllers/AnnouncementExampleViewController.swift]: :eyes: `let`

File: Example/Sources/View Controllers/AnnouncementExampleViewController.swift [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+struct MediaItemImpl: MediaItem {
+    var url: URL?
+    var image: UIImage?
+    var placeholderImage: UIImage
+    var size: CGSize
+}
+
+struct User: SenderType {
+    var senderId: String
+    var displayName: String
```
Comment (manicmaniac) [inline Example/Sources/View Controllers/AnnouncementExampleViewController.swift]: :eyes: `let`

File: Example/Sources/View Controllers/AnnouncementExampleViewController.swift [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+
+    static func me() -> User {
+        return User(senderId: "me", displayName: "my user")
```
Comment (manicmaniac) [inline Example/Sources/View Controllers/AnnouncementExampleViewController.swift]: Better to declare as `static let`.

File: Sources/Layout/AnnounceMessageSizeCalculator.swift [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    
+    open override func avatarPosition(for message: MessageType) -> AvatarPosition {
+        return avatarPosition
+    }
+    
+}
+
+fileprivate extension UIEdgeInsets {
+    init(top: CGFloat = 0, bottom: CGFloat = 0, left: CGFloat = 0, right: CGFloat = 0) {
+        self.init(top: top, left: left, bottom: bottom, right: right)
```
Comment (manicmaniac) [inline Sources/Layout/AnnounceMessageSizeCalculator.swift]: Really needed?
A user of this initializer should know about the default values.

File: Sources/Models/AvatarPosition.swift [hunk 11]
```diff
@@ -64,6 +68,8 @@ public struct AvatarPosition: Equatable {
         /// Aligns the `AvatarView`'s bottom edge to the cell's bottom edge.
         case cellBottom
         
+        /// Positions when the `AvatarView` isn't needed.
+        case none
```
Comment (manicmaniac) [inline Sources/Models/AvatarPosition.swift:72]: :+1: for a descriptive comment.

File: Sources/Views/MessageContainerView.swift [hunk 12]
```diff
@@ -28,61 +28,44 @@ open class MessageContainerView: UIImageView {
...
-            mask = nil
-            image = nil
-            tintColor = nil
-            configurationClosure(self)
+        case .leftBubble:
+            roundingCorners = [.topRight, .bottomLeft, .bottomRight]
+        case .rightBubble:
+            roundingCorners = [.topLeft, .bottomLeft, .bottomRight]
+        default:
+            fatalError("This style never call this function.")
```
Comment (manicmaniac) [inline Sources/Views/MessageContainerView.swift]: `s/call/calls/`

**Comment by m-sugawara** [comment]
Thanks a lot!! I will check your comments.

File: Example/Sources/View Controllers/AnnouncementExampleViewController.swift (hunk 3 above)
Comment (m-sugawara) [reply Example/Sources/View Controllers/AnnouncementExampleViewController.swift]: Ah, I want to leave these debug prints. Because these are examples.

Comment (manicmaniac) [reply Example/Sources/View Controllers/AnnouncementExampleViewController.swift]: Thanks, got it :+1:

**Comment by m-sugawara** [comment]
@manicmaniac 
I fixed codes pointed out. PTAL~ 🙏

**Review by manicmaniac — APPROVED** [review]
LGTM :+1:

**Comment by m-sugawara** [comment]
Thank you so much!! 🙏🙏

**Pull request merged**